### ISSUE TITLE
Stop overriding the original app

### ISSUE THREAD

**Issue by xavierchow** [Wiredcraft/api-boilerplate#37]
There is a lib/app instead of original loopback app, it's so confusing, and the merit of the overriding is for constructing restAPI with configs which I think is not needed anymore.

cc @sanagnos @paulvollmer @makara @ChopperLee2011

**Comment by makara** [comment]
The main purpose was to get a singleton, but yeah when there's the convenience people tend to abuse. I don't really know how to solve it yet, maybe I can change the lifecycle, but then you'll need to pass it with the arguments all the time and it can be annoying. Any ideas? We can talk.

**Comment by xavierchow** [comment]
FYI, https://github.com/Wiredcraft/sams-customer-support/issues/32,
This is not the first time somebody got trapped by this kind of confusion(including me).

**Comment by makara** [comment]
Commented in there. It's exactly what I talked about, people can use `app` all the time so people are using it everywhere. For the case I'd build another config thing that doesn't require boot. Let me try something.

**Comment by xavierchow** [comment]
@pivstone could you send a PR to remove the overridden app, just same as you did for user-preference.

**Comment by pivstone** [comment]
sure.

**Comment by makara** [comment]
One thing comes with the extended lib is the ability to "close" or "reboot" the app, so we can use a singleton to do the test.

So if you don't want a singleton, then don't. You should boot a new app instance with each of the tests, or say every time you need a booted app. Furthermore, `boot` is async, so simply require the `server` file doesn't always work, and you are doing it wrong with removing all the `before` from the tests.

**Comment by makara** [comment]
> 10.health.test.js do not access app.get()
> supertest needs the server.js to boot up the test server

If you don't need a booted app, sure, but what does supertest need here? the middleware right? the app needs to be booted to load the middlewares right?

> the PR's goal is remove app.js file

You want to remove it because you want to remove it? or because someone asked you to?

**Comment by pivstone** [comment]
> If you don't need a booted app, sure, but what does supertest need here? the middleware right? the app needs to be booted to load the middlewares right?

Actually, `supertest` does NOT inject middleware. Ref the source code, `supertest`  use http.createServer to create a test server with a random port. Then use `superagent` send a request to the test server.

I don't have any experience in NodeJs, but if you can help me learn more about anything in Nodejs I will very happy.

> You want to remove it because you want to remove it? or because someone asked you to?

I understand we haven't reached the solution to this issue, so I closed the RP in this morning.  Anyway thanks for your review.

Refs:
  
- https://nodejs.org/api/http.html#http_http_createserver_options_requestlistener
- https://github.com/visionmedia/supertest/blob/master/index.js#L19-L21

**Comment by makara** [comment]
Hmm, my suggestion was to create and boot a new app instance with every test, not to close the PR, but sure you guys decide.

**Comment by xavierchow** [comment]
- back to the beginning, the test code should never use the `app.get()` coz it's an internal logic, test code should hit the API externally.
- supertest/superagent will create an app instance everytime you pass an `app` to it.
- we don't need reboot for each test case.
- about the singleton, `require('../server.js')` is already singleton, coz commonJS caches the required module.

Rebuilding a lib/app to tweak the boot seems to be a black magic which is not necessary, instead of cutomizing the app more important thing is writing the test code in the right way.

**Comment by makara** [comment]
> we don't need reboot for each test case

Again, `boot` is async, so your example may or may not work - to me it's all by luck, unless you can prove `request(server)` always wait for the `boot`.

Furthermore, what if you want to change some env var for a certain test?

**Comment by xavierchow** [comment]
Yes, `boot` is async and `supertest.request` doesn't wait for it, but let's be pragmatic, in my example the boot happens in the load time(CommonJS require) and the `supertest.request` happens in the runtime `describe.it`, the interval should be large enough to let the boot finishes.

Also, you don't need to build another app instance even if you want to manipulate the `boot`, you can export the boot handling,
```
server.js
boot(app, __dirname, function(err) {

**Comment by xavierchow** [comment]
done, closing

**Comment by CCharlieLi** [comment]
are we abandon `loopback-extended-lib`? since I saw only `file-register` is used [here](https://github.com/Wiredcraft/api-boilerplate/blob/master/server/lib/index.js#L7)

And let's also clean up the code since there's no more lib.app
https://github.com/Wiredcraft/api-boilerplate/blob/master/server/lib/debug.js#L6
https://github.com/Wiredcraft/api-boilerplate/blob/master/server/lib/loggers/syslog.js#L6

**Comment by xavierchow** [comment]
yes, loopback-extended-lib is not needed, `require` from CommonJS is good enough.

**Comment by xavierchow** [comment]
Created https://github.com/Wiredcraft/api-boilerplate/pull/56 for it.
